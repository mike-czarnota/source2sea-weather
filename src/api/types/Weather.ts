export interface IWeatherResponse {
  main: {
    humidity: number;
    temp: number;
    temp_max: number;
    temp_min: number;
  };
  sys: {
    sunrise: number;
    sunset: number;
  };
  weather: {
    main: string;
  }[];
  visibility: number;
  name: string;
}
