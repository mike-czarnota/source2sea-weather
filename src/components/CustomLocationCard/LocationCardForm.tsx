import { ChangeEvent, FormEvent, useEffect, useRef, useState } from 'react';

export type TSubmitValues = { locationName: string };

interface ILocationCardInputProps {
  onSubmit: (values: TSubmitValues) => void;
}

const LocationCardForm: React.FC<ILocationCardInputProps> = ({ onSubmit }) => {
  const [userLocation, setUserLocation] = useState('');
  const [error, setError] = useState('');

  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    inputRef.current?.focus();
  }, []);

  const onFormSubmit = (event: FormEvent) => {
    event.preventDefault();

    if (userLocation) {
      onSubmit({
        locationName: userLocation,
      });
    } else {
      setError('Please, provide a location');
    }
  };

  const onInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    setError('');
    setUserLocation(event.target.value);
  };

  return (
    <div className="flex flex-col">
      <form onSubmit={onFormSubmit}>
        <input
          ref={inputRef}
          className="w-full"
          type="text"
          placeholder="Where do you want to check the weather?"
          value={userLocation}
          onChange={onInputChange}
        />
      </form>

      {error && <span className="text-red-400">{error}</span>}

      <span className="text-sm italic text-gray-400">
        Hit &quot;Enter&quot; to search
      </span>
    </div>
  );
};

export default LocationCardForm;
