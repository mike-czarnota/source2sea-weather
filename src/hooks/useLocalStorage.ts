import { useCallback } from 'react';

type TGetItem = <TReturnType>() => TReturnType;
type TSetItemValue = any;
type TSetItem = (value: TSetItemValue) => void;

export const useLocalStorage = (
  key: string,
  defaultValue: any
): [TGetItem, TSetItem] => {
  const setItem: TSetItem = useCallback(
    (value) => {
      localStorage.setItem(key, JSON.stringify(value));
    },
    [key]
  );

  const getItem: TGetItem = useCallback(() => {
    const item = localStorage.getItem(key);
    if (!item) {
      setItem(defaultValue);
      return defaultValue;
    }

    return JSON.parse(item);
  }, [key, defaultValue]);

  return [getItem, setItem];
};
