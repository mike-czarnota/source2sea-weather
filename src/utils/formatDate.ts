import format from 'date-fns/format';

export const formatTime = (date: Date): string => format(date, 'HH:mm');
