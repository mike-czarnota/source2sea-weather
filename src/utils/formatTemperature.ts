export const formatTemperature = (temperature: number): string =>
  `${temperature}Â°C`;
