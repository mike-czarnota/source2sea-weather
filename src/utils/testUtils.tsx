import { render, RenderResult } from '@testing-library/react';

import { BrowserRouter } from 'react-router-dom';

export const renderWithWrapper = (ui: React.ReactElement): RenderResult =>
  render(<BrowserRouter>{ui}</BrowserRouter>);
