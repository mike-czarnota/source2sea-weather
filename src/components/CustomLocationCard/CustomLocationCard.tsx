import { useEffect, useRef, useState } from 'react';
import { LocationCardContainer } from '../LocationCard';
import LocationCardForm, { TSubmitValues } from './LocationCardForm';

interface ICustomLocationCardProps {}

const CustomLocationCard: React.FC<ICustomLocationCardProps> = () => {
  const [isEditing, setIsEditing] = useState(false);

  const userLocationRef = useRef('');

  useEffect(() => {
    if (!userLocationRef.current) {
      setIsEditing(true);
    }
  }, []);

  const toggleEdit = () => {
    setIsEditing(true);
  };

  const onSubmit = (values: TSubmitValues) => {
    userLocationRef.current = values.locationName;
    setIsEditing(false);
  };

  return (
    <div>
      {isEditing ? (
        <LocationCardForm onSubmit={onSubmit} />
      ) : (
        <>
          <LocationCardContainer locationName={userLocationRef.current} />

          <button type="button" className="text-blue-500" onClick={toggleEdit}>
            Edit
          </button>
        </>
      )}
    </div>
  );
};

export default CustomLocationCard;
