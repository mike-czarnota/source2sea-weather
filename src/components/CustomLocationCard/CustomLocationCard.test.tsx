import pretty from 'pretty';
import { act, screen } from '@testing-library/react';
import { Client, ClientContextProvider } from 'react-fetching-library';
import { renderWithWrapper } from '../../utils/testUtils';
import CustomLocationCard from './CustomLocationCard';
import { IWeatherResponse } from '../../api/types/Weather';
import { useLocalStorage } from '../../hooks/useLocalStorage';

jest.mock('../../hooks/useLocalStorage', () => ({
  useLocalStorage: jest.fn(),
}));

describe('testing CustomLocationCard.test.tsx', () => {
  it('should fetch data and render correctly', async () => {
    (useLocalStorage as unknown as jest.Mock).mockImplementation(() => {
      const getItem = () => {
        return {
          payload: {
            weather: [
              {
                main: 'Sunny',
              },
            ],
            name: 'Kuala Lumpur',
            main: {
              temp: 900,
              humidity: 901,
              temp_max: 902,
              temp_min: 899,
            },
            sys: {
              sunrise: Date.now(),
              sunset: Date.now(),
            },
            visibility: 2390,
          },
        };
      };

      const setItem = () => {};

      return [getItem, setItem];
    });

    const payload: IWeatherResponse = {
      weather: [
        {
          main: 'Sunny',
        },
      ],
      name: 'Kuala Lumpur',
      main: {
        temp: 100,
        humidity: 101,
        temp_max: 102,
        temp_min: 99,
      },
      sys: {
        sunrise: Date.now(),
        sunset: Date.now(),
      },
      visibility: 2390,
    };

    const client = {
      query: async () => ({
        error: false,
        status: 200,
        payload,
      }),
    };

    jest.useFakeTimers();

    const component = renderWithWrapper(
      <ClientContextProvider client={client as unknown as Client}>
        <CustomLocationCard />
      </ClientContextProvider>
    );

    expect(pretty(component.baseElement.innerHTML)).toMatchSnapshot();

    expect(screen.getByText('Kuala Lumpur')).toBeInTheDocument();
    expect(screen.getByText('900°C')).toBeInTheDocument();
    expect(screen.getByText('Edit')).toBeInTheDocument();

    await act(async () => {
      jest.runAllTimers();
    });

    expect(pretty(component.baseElement.innerHTML)).toMatchSnapshot();

    expect(screen.getByText('Kuala Lumpur')).toBeInTheDocument();
    expect(screen.getByText('100°C')).toBeInTheDocument();
    expect(screen.getByText('Edit')).toBeInTheDocument();
  });

  it('should not load when empty', async () => {
    (useLocalStorage as unknown as jest.Mock).mockImplementation(() => {
      const getItem = () => {
        return '';
      };

      const setItem = () => {};

      return [getItem, setItem];
    });

    const payload: IWeatherResponse = {
      weather: [
        {
          main: 'Sunny',
        },
      ],
      name: 'Kuala Lumpur',
      main: {
        temp: 100,
        humidity: 101,
        temp_max: 102,
        temp_min: 99,
      },
      sys: {
        sunrise: Date.now(),
        sunset: Date.now(),
      },
      visibility: 2390,
    };

    const client = {
      query: async () => ({
        error: false,
        status: 200,
        payload,
      }),
    };

    const component = renderWithWrapper(
      <ClientContextProvider client={client as unknown as Client}>
        <CustomLocationCard />
      </ClientContextProvider>
    );

    expect(pretty(component.baseElement.innerHTML)).toMatchSnapshot();

    expect(
      screen.getByPlaceholderText('Where do you want to check the weather?')
    ).toBeInTheDocument();
    expect(screen.getByText('Hit "Enter" to search')).toBeInTheDocument();
  });

  it('should load with error', async () => {
    (useLocalStorage as unknown as jest.Mock).mockImplementation(() => {
      const getItem = () => {
        return {
          payload: {
            name: 'Salt Lake City',
          },
        };
      };

      const setItem = () => {};

      return [getItem, setItem];
    });

    const payload: IWeatherResponse = {
      whatever: false,
    } as unknown as IWeatherResponse;

    const client = {
      query: async () => ({
        error: true,
        status: 200,
        payload,
      }),
    };

    jest.useFakeTimers();

    const component = renderWithWrapper(
      <ClientContextProvider client={client as unknown as Client}>
        <CustomLocationCard />
      </ClientContextProvider>
    );

    expect(pretty(component.baseElement.innerHTML)).toMatchSnapshot();
    expect(
      screen.getByText('Error while loading location:')
    ).toBeInTheDocument();
    expect(screen.getByText('Salt Lake City')).toBeInTheDocument();
    expect(screen.getByText('Edit')).toBeInTheDocument();

    await act(async () => {
      jest.runAllTimers();
    });
  });
});
