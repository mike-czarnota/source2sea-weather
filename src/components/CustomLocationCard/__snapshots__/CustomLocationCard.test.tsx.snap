// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing CustomLocationCard.test.tsx should fetch data and render correctly 1`] = `
"<div>
  <div><a class=\\"self p-5 md:px-10 md:py-7 flex justify-between items-center border rounded-lg bg-white border-gray-300 hover:border-gray-900 duration-200 ease-in-out shadow-md hover:shadow-2xl\\" href=\\"/Kuala Lumpur\\"><span class=\\"text-2xl\\">Kuala Lumpur</span><span class=\\"text-xl\\">900°C</span></a><button type=\\"button\\" class=\\"text-blue-500\\">Edit</button></div>
</div>"
`;

exports[`testing CustomLocationCard.test.tsx should fetch data and render correctly 2`] = `
"<div>
  <div><a class=\\"self p-5 md:px-10 md:py-7 flex justify-between items-center border rounded-lg bg-white border-gray-300 hover:border-gray-900 duration-200 ease-in-out shadow-md hover:shadow-2xl\\" href=\\"/Kuala Lumpur\\"><span class=\\"text-2xl\\">Kuala Lumpur</span><span class=\\"text-xl\\">100°C</span></a><button type=\\"button\\" class=\\"text-blue-500\\">Edit</button></div>
</div>"
`;

exports[`testing CustomLocationCard.test.tsx should load with error 1`] = `
"<div>
  <div>
    <div class=\\"italic\\">Error while loading location: <span class=\\"font-bold\\">Salt Lake City</span></div><button type=\\"button\\" class=\\"text-blue-500\\">Edit</button>
  </div>
</div>"
`;

exports[`testing CustomLocationCard.test.tsx should not load when empty 1`] = `
"<div>
  <div>
    <div class=\\"flex flex-col\\">
      <form><input class=\\"w-full\\" type=\\"text\\" placeholder=\\"Where do you want to check the weather?\\" value=\\"\\"></form><span class=\\"text-sm italic text-gray-400\\">Hit \\"Enter\\" to search</span>
    </div>
  </div>
</div>"
`;
